jetty-ee9/jetty-ee9-webapp/src/test/java/org/eclipse/jetty/ee9/webapp/WebAppContextTest.java

```diff
@@ -546,6 +551,132 @@ public void testBaseResourceAbsolutePath() throws Exception
         assertTrue(context.isAvailable(), "WebAppContext should be available");
     }
 
+    @Test
+    public void testGetResourceFromCollection() throws Exception
+    {
+        Server server = newServer();
+
+        WebAppContext context = new WebAppContext();
+        context.setContextPath("/");
+        context.setBaseResource(ResourceFactory.combine(
+            ResourceFactory.root().newResource(MavenTestingUtils.getTestResourcePath("wars/layer0/")),
+            ResourceFactory.root().newResource(MavenTestingUtils.getTestResourcePath("wars/layer1/"))));
+        server.setHandler(context);
+        server.start();
+
+        ServletContext servletContext = context.getServletContext();
+        assertThat(servletContext.getResource("/WEB-INF/zero.xml"), notNullValue());
+        assertThat(servletContext.getResource("/WEB-INF/one.xml"), notNullValue());
+    }
+
+    @Test
+    public void testGetResourcePathsFromCollection() throws Exception
+    {
+        Server server = newServer();
+
+        WebAppContext context = new WebAppContext();
+        context.setContextPath("/");
+        context.setBaseResource(ResourceFactory.combine(
+            ResourceFactory.root().newResource(MavenTestingUtils.getTestResourcePath("wars/layer0/")),
+            ResourceFactory.root().newResource(MavenTestingUtils.getTestResourcePath("wars/layer1/"))));
+        server.setHandler(context);
+        server.start();
+
+        ServletContext servletContext = context.getServletContext();
+        assertThat(servletContext.getResourcePaths("/WEB-INF/"), containsInAnyOrder("/WEB-INF/zero.xml", "/WEB-INF/one.xml"));
+    }
+
+    @Test
+    public void testGetResourcePathsWithDirsFromCollection() throws Exception
+    {
+        Server server = newServer();
+
+        WebAppContext context = new WebAppContext();
+        context.setContextPath("/");
+        context.setBaseResource(ResourceFactory.combine(
+            ResourceFactory.root().newResource(MavenTestingUtils.getTestResourcePath("wars/layer0/")),
+            ResourceFactory.root().newResource(MavenTestingUtils.getTestResourcePath("wars/layer1/")),
+            ResourceFactory.root().newResource(MavenTestingUtils.getTestResourcePath("wars/with_dirs/"))
+        ));
+        server.setHandler(context);
+        server.start();
+
+        ServletContext servletContext = context.getServletContext();
+        Set<String> results = servletContext.getResourcePaths("/WEB-INF/");
+        String[] expected = {
+            "/WEB-INF/zero.xml",
+            "/WEB-INF/one.xml",
+            "/WEB-INF/bar/",
+            "/WEB-INF/foo/"
+        };
+        assertThat(results, containsInAnyOrder(expected));
+    }
+
+    @Test
+    @Disabled("There is extra decoding of the nested-reserved paths that is getting in the way")
```

Comment:
This can be fixed with @lorban backporting his work in PR #9974 back to ee9 tree.

Reply:
That would also require backporting the fix to 10/11.